### PR TITLE
Goyalbl code review 2

### PR DESCRIPTION
**Pull request by bipalgoyal**
**Analysis of the program.**
Cookit is an android application that one uses for cooking. It allows users to find and create recipes based on several factors like time, ingredients, etc.

**Was the program available in UC Github on time?**
Yes, I had enough time to complete the code review for this project.

**Is the program easy for you, as an outsider, to understand?**
Yes, it was straightforward to envision the application in my head after looking at the interactive storyboard and the detailed class diagram of the project. 

**Does the program compile?**
Yes, the program compiles

I have proposed the following changes, which I will mention here and then describe in detail if needed inside the file changes:

#### 1. Modified one of the Integration test to use the let scope function 

Their integration test uses not-null assertion operator (!!), which might throw an exception if a null object is passed to it. It is instead better to use the let scope function which doens't throw a null pointer exception. 

**Source(s):**
- Lecture taught by Brandan Jones in the classroom
- https://czerwinski.it/2018/11/30/the-case-against-double-exclamation-operator.html 

#### 2. Wrote unit tests for each DTO in the application
I created some unit tests in order to test the DTO objects that were being used in the application

**Source(s):** 
https://fortegrp.com/the-importance-of-unit-testing

#### 3. Split one big file containing multiple classes to separate files for each class
The Recipe.kt file contained code for 2 different DTOs. I split them into multiple files, which is generally considered best practice.

**Source(s)**
https://softwareengineering.stackexchange.com/questions/57174/which-is-better-many-class-definitions-in-the-same-file-or-every-class-definiti

#### Link to my commits:
I have more than three, I just have links to random ones:
1. https://github.com/pranavm7/UniConnect/commit/bc89d1d2c64778201b26d244c3f232180a592fa7
2. https://github.com/pranavm7/UniConnect/commit/39e7c6eaa7d486adb0d728ebd8e341ac56956bba
3. https://github.com/pranavm7/UniConnect/commit/9205b90cef018d5087aa6f7ad89940d7b7e199a3

#### What did I learn over here?
1. I learned much more about how scope functions work in Kotlin.
4. I learned the importance of Unit Testing and its benefits.
5. I learned about the benefits of separating code properly in multiple files for a big application.

### CODE REVIEW

**Comment by ccbrowndev** [comment]
To give some reasoning behind RecipeList being in the Recipe.kt file:
* It is very closely related to the Recipe data class
* It is there to only to facilitate parsing of Recipe objects

Moving it to a separate file may be a good idea, but that was our reasoning for including them in the same file.

That being said I find it confusing that you thought to separate RecipeList class into a separate file, but put Meal unit tests into the Recipe unit test file.

**Comment by bipalgoyal** [comment]
Now that I hear your reasoning, I understand why you kept both of those classes on the same file. However, it honestly confused me when I was going through the program and saw references to RecipeList, but couldn't find it's file on the project tab. I believe that it might confuse others as well, and think it best to separate them.

Also, regarding the unit test, that is certainly an error on my part. I didn't realize that you had separate unit test files for recipes and meals. I would probably now split them into multiple files. 

Cheers!

File: app/src/main/java/com/cookit/dto/RecipeList.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.cookit.dto
+
+import com.google.gson.annotations.SerializedName
+
+data class RecipeList(@SerializedName("meals") var recipes : ArrayList<Recipe> = arrayListOf()) {}
```
Comment (discospiff) [inline app/src/main/java/com/cookit/dto/RecipeList.kt:5]: It makes sense for this to be in its own file.

File: app/src/test/java/com/cookit/RecipeIntegrationTest.kt [hunk 2]
```diff
@@ -51,12 +51,13 @@ class RecipeIntegrationTest {
...
         assertTrue(allRecipes!!.recipes.isNotEmpty())
         var containsAppleFrangipanTart = false
-        allRecipes!!.recipes.forEach{
-            if (it.name.lowercase().contains("apple") && it.cuisine == "British") {
-                containsAppleFrangipanTart = true
+        allRecipes?.let {
+            it.recipes.forEach {
+                if (it.name.lowercase().contains("apple") && it.cuisine == "British") {
+                    containsAppleFrangipanTart = true
+                }
```
Comment (discospiff) [inline app/src/test/java/com/cookit/RecipeIntegrationTest.kt:58]: Much better.  let{} is an improvement over !!

File: app/src/test/java/com/cookit/RecipeUnitTest.kt [hunk 3]
```diff
@@ -45,6 +43,22 @@ class RecipeUnitTest {
...
+        Assert.assertTrue(meal.description.equals("A tasty snack"))
+    }
+
+    @Test
+    fun `Given a recipe DTO when name is Espresso and category is coffee`(){
+        val recipe = Recipe("","Espresso", "coffee")
+        Assert.assertTrue(recipe.name.equals("Espresso"))
+        Assert.assertTrue(recipe.category.equals("coffee"))
+    }
+
```
Comment (discospiff) [inline app/src/test/java/com/cookit/RecipeUnitTest.kt:59]: More test coverage is a good idea.

**Comment by discospiff** [comment]
Several good suggestions here that reduce technical debt.
I recommend merging this branch.

**Pull request merged**